eeps/eep-0058.md

```diff
@@ -0,0 +1,255 @@
+
+
+Map generator should be semantically equivalent to the list generator
+running over the result of `maps:to_list/1`, when the input is a
+map datatype:
+
+    #{ ... || K := V <- Map}
+
+is equivalent to
+
+    #{ ... || {K, V} <- maps:to_list(Map)}
+
+And for `maps:iterator()` as input - it can't be easily expressed as
+equivalent list generator. But the idea is that it's equivalent to
+the consumption of the iterator by repetitive calls to `maps:next/1`
+and matching `Key` and `Value` over key and value patterns
+accordingly on each iteration loop.
+
+Map generator should raise `error({bad_generator, ExprResult})`
+if the `MapOrIterator` expression evaluation produced anything but
+map or map iterator.
+
+Key pattern is allowed to be catch-all / contain unbound variables
+(which is not allowed in other map pattern-matching contexts).
+
+Variable binding and shadowing rules should be the same as for list
+and binary comprehensions (which follows from the equivalence rules
+described above).
+
+Rationale
+=========
+
+### Syntax for arrow, generator pattern and comprehension key-value pair
+
+We decided to go with the syntax proposed in EEP-43.
+
+`#{ .. }` were chosen for the outer tokens of map comprehension
+because the same are used when constructing a new map - rule similar to
+list and binary comprehensions.
+
+`K := V <-` is not ambiguous for the parser, so unlike binary
+generator, there is no need for a new arrow. `<-` was chosen over
+`<=` because the latter is less visually distinguishable when it is
+placed relatively close to the `=>` and `:=` tokens:
+
+    #{K => V || K := V <= Map}.
+    % vs
+    #{K => V || K := V <- Map}.
+
+`KeyPattern := ValuePattern` is the same syntax that is used in
+map pattern-matching (however, in map generators KeyPattern is also
+allowed to be a match-all / unbound variable).
+
+`KeyExpr => ValueExpr` is the same syntax that is used for the new
+map construction.
+
+### Order in which map generator yields key-value pairs
+
+Since maps do not have any order specified, the order in which the
+generator would yield key-value pairs should not be guaranteed (same
+way as the ordering of the list, resulting from `maps:to_list/1` is
+not guaranteed).
+
+### Behavior when comprehension produced duplicate keys
+
+Same as `maps:from_list/1`: the latest (right-most) value is used
+and the previous values are ignored. It might be less predictable
+when a map comprehension is used in combination with a map generator
+(because the map generator order is not defined), but is important when
+a map comprehension is used with a list or a binary generator.
+
+### Whether to allow map iterator as input for map generator
+
+Map iterators are allowed in `maps:filter/2`, `maps:fold/3` and
+`maps:map/2`. Since comprehensions are often used in the same
+contexts as those functions, it makes sense to allow map iterators
+so comprehensions could be used as a drop-in replacement for the
+aforementioned functions.
```

Comment:
I'm still in doubt though... `maps:iterator()` is part of a standard library, but here we making it part of the language? Do we have another examples where this happens?

Reply:
True... I at least don't know of any other place where something like this is done.

Reply:
Neither do I.